### ISSUE TITLE
"The recipient's key is not valid" errors

### ISSUE THREAD

**Issue by charlesmchen** [signalapp/Signal-iOS#1838]

**Comment by charlesmchen** [comment]
@michaelkirk Unfortunately I can no longer reproduce this issue. Closing for now; will reopen if this issue re-emerges.

**Comment by riyapenn** [comment]
Error at 21 March 2017 at 17:45 and 17:22 
https://gist.github.com/anonymous/95c1f42354bb255e7b5afdc8007e003e

**Comment by charlesmchen** [comment]
@riyapenn I'm working on this issue already. I've repro'd twice in the last
two months.  I've got an open PR that adds logging that will help us
diagnose this.  Every time you see this (including the time for the logs
above) please document the following:

* Was this 1:1 or group convo?
* Did sender or receiver see these errors?
* Platform / version of sender _and_ receiver.
* Had either party recently woken up?
* Had either party not been used in a while?
* Had either reinstalled since they last communicated?
* Did you see "safety numbers changed" ui right before these errors?  Did
these errors show up before or after you approved the new safety numbers?
Were there more than one "safety numbers changed" message?
* Was there any other interaction on either party's part before these
errors?
* How many times did this error show up in the conversation?

**Comment by precipice** [comment]
Just saw this; my key was reported as invalid for the recipient.

* 1:1
* Receiver (not me)
* Sender: iOS 10.3, Receiver: iOS 10.3
* Not sure what you mean by woken up.
* We had not sent signal messages to each other for some time.
* I have reinstalled since we last communicated. Twice, actually.
* The receiver did see "Safety numbers changed" before the errors started. They turned into those errors after the receiver accepted the new safety numbers.
* The initial contact between us this time was me trying to initiate a call to the receiver, three times without success.
* The error appeared for the receiver eight times.

**Comment by charlesmchen** [comment]
Thanks for the info @precipice .  We're working on this.

**Comment by michaelkirk** [comment]
@precipice It would be helpful if you (and ideally your conversation partner) could provide debug logs  and point us to the corresponding clock time when you saw the behavior.

To submit a log: Settings > Advanced > Submit debug logs

Thanks for any help!

**Comment by precipice** [comment]
@michaelkirk I had debug logs off at the time. I'll try to reproduce the error now with logs on, and will ask the recipient to submit logs as well.

**Comment by precipice** [comment]
@michaelkirk well, I turned on debug logs and sent another message, and the recipient got it. My favorite kind of bug!

The recipient submitted debug logs just now. The first call attempt was at 2:34pm US/Pacific time, today. There were failed text messages at 3:19pm US/Pacific (I sent one to the recipient, which did produce the error, and they sent one back to me, which did not--I was able to read their message). The successful messages were at 3:49pm US/Pacific (I sent him one he received, and vice versa).

**Comment by timothyzillion** [comment]
I was on the other side of the the conversation Mr. Precipice mentioned. I did have debug logs enabled and submitted them:
https://gist.github.com/anonymous/7daf1d31c5f1d5bf51de31cf4ae5e4f0

**Comment by wiseman** [comment]
I just ran into this also.

* Was this 1:1 or group convo?
1:1

* Did sender or receiver see these errors?

Sender.  Receiver saw no errors.

* Platform / version of sender _and_ receiver.

Sender: 2.9.0, Receiver: Unknown.

* Had either party recently woken up?

Unknown.

* Had either party not been used in a while?

Neither side had used SIgnal recently (weeks, at least).

* Had either reinstalled since they last communicated?

Yes, both.

* Did you see "safety numbers changed" ui right before these errors?  Did
these errors show up before or after you approved the new safety numbers?

Yes, I (sender) saw "safety numbers changed".  When I tapped verify -> show safety numbers, the app crashed--repeatedly.  The invalid key error only showed up immediately after I accepted the new safety numbers.

Were there more than one "safety numbers changed" message?

No.

* Was there any other interaction on either party's part before these
errors?

No.

* How many times did this error show up in the conversation?

I got the invalid key error, deleted and re-added the contact, tried to send some messages that did not go through, then got the invalid key error again, then one of my messages got through. At that point, we were able to send messages back and forth successfully.

**Comment by riyapenn** [comment]
(This information is separate from the debug log listed in https://github.com/WhisperSystems/Signal-iOS/issues/1838#issuecomment-288251985)  

* Was this 1:1 or group convo?
1:1 conversation 

* Did sender or receiver see these errors?
Receiver saw this error 

* Platform / version of sender _and_ receiver
Version of receiver:2.9.0 ; Version of sender: latest version (just downloaded)

* Had either party recently woken up?
If by recently woken up you mean being used for the first time in a little while? Yes on both sender and receivers part

* Had either party not been used in a while?
Neither party has used since Jan 15 2017

* Had either reinstalled since they last communicated?
Neither has been reinstalled since the error occurred

* Did you see "safety numbers changed" ui right before these errors?  Did
these errors show up before or after you approved the new safety numbers?
Were there more than one "safety numbers changed" message?
Saw safety number changed, went to click verify the change, then the error occurred

* Was there any other interaction on either party's part before these
errors?
No previous interaction, these were the first interaction since 1/15/2017

* How many times did this error show up in the conversation?
Only once, was able to message each other after the error occurred

Debug Logs:
Receiver: https://gist.github.com/ef1d058e354e67c1818de3d00cb78bb9
Sender: N/A yet

**Comment by Be-ing** [comment]
* Was this 1:1 or group convo?

1:1

* Did sender or receiver see these errors?

receiver

* Platform / version of sender _and_ receiver.

Sender: Android (me)
Receiver: iOS

* Had either party recently woken up?

I don't know what you mean by "woken up".

* Had either party not been used in a while?

We both use Signal pretty regularly.

* Had either reinstalled since they last communicated?

Yes, my Android phone was stolen. Before I got a new phone, I sent a message with Signal Desktop, which was received and shown correctly. When I got a new Android phone, messages that I sent from the new phone showed the error for my friend on iOS.

* Did you see "safety numbers changed" ui right before these errors? Did these errors show up before or after you approved the new safety numbers? Were there more than one "safety numbers changed" message?

There was no warning about safety numbers for either of us.

* Was there any other interaction on either party's part before these errors?

yes

* How many times did this error show up in the conversation?

It showed up for each message until the receiver who saw the error sent me a message. Thereafter either of us could send messages.

**Comment by michaelkirk** [comment]
Hi @Be-ing can you submit a debug log? And can you ask your conversation partner to submit one as well?

And then, so we can track down the correct log entries, about what time did this happen?

**Comment by michaelkirk** [comment]
Sorry to nag @Be-ing, but the debug logs self destruct after three days. 

Is it possible that you could submit one here and, even better, ask your conversation partner to also submit one here?

**Comment by pde** [comment]
Here are some more logs of this problem:

https://gist.github.com/anonymous/b19a8bf382b3c97e8dec98035b30d97c

The errors in question would be 18:04 onwards at the end of the log, and perhaps also around 11:42 yesterday. In both cases these are iOS devices receiving the message; I also have Signal desktop running without problems. In at least one of those cases (11:42 yesterday) had to accept a new key on both clients but then got the invalid message on iOS after doing so. I suspect the iOS device was asleep at the start of both conversations, but would no longer have been asleep by the time of the later messages in the 18:04 conversation today.

**Comment by pde** [comment]
Interestingly in the 11:42 case yesterday, the iOS UI showed the "The receipient's key is not valid" message twice, even though I'd only been sent one message.

@michaelkirk ping me if screenshots of the way this showed up in app would be useful.

**Comment by michaelkirk** [comment]
Thanks for the logs @pde. Any way you could get the logs from the other person? Often it's helpful to get both sides when trying to put the picture together.

**Comment by pde** [comment]
@michaelkirk ugh sorry I just saw this :(.  Presuming it's too late to get them since they're three-day-rotating, but I'll do so if I see the bug again.

**Comment by tommi** [comment]
@michaelkirk 


* Was this 1:1 or group convo? => 1:1
* Did sender or receiver see these errors? => Receiver
* Platform / version of sender _and_ receiver. => Both iOS 2.11.2
* Had either party recently woken up? => ?
* Had either party not been used in a while? => not in active use
* Had either reinstalled since they last communicated? Sender restored from backup to new phone
* Did you see "safety numbers changed" ui right before these errors? => Receiver: yes
* Did these errors show up before or after you approved the new safety numbers? => Receiver: after
* Were there more than one "safety numbers changed" message? => Receiver: no
* Was there any other interaction on either party's part before these errors? => a discussion long time ago
* How many times did this error show up in the conversation? => Twice, once per sent message

Sender: https://gist.github.com/anonymous/492efebdb0b06304e614d1820e20857a
Receiver: https://gist.github.com/anonymous/51021f3620449e775270f5701f73b72e

**Comment by tommi** [comment]
Deleting sender account and reregistering it caused another "safety numbers changed" error on the receiver, but this time accepting it made the conversation work.

**Comment by michaelkirk** [comment]
sorry @tommi, and thanks for the good logs.

Looking at the logs it seems this was an issue on the recipient side fixed in 2.11. Even though the recipient was running an up-to-date versions when they *saw* the error, the root cause occurred earlier when the recipient was running a pre 2.11 version.

Please let us know if you see this again.

**Comment by riyapenn** [comment]
* Was this 1:1 or group convo? - 1 to 1

* Did sender or receiver see these errors? - No.

* Platform / version of sender _and_ receiver. - Mine is the OS X app (& then iOS latter that I noticed was giving me this error) his is just iOS.  My iOS still works w/ everyone else, just not this one particular person.

* Had either party recently woken up? -I’m not sure what you mean here, but my OS X app was running at the time it started.

* Had either party not been used in a while? - I don’t have a lot of messages come thru Signal, so it was the first in probably 2 weeks.

* Had either reinstalled since they last communicated? - No.

* Did you see "safety numbers changed" ui right before these errors? Did these errors show up before or after you approved the new safety numbers? Were there more than one "safety numbers changed" message? - Just one message. I don’t remember exactly the wording of the message, it didn’t strike me as something unusual at the time.  I never receive the message before on any platform, and never saw it again to date.

* Was there any other interaction on either party's part before these errors? -No he messaged me, my desktop app popped up that my ke needed to be updated, I clicked to update it, the message came thru.  Only later did I realize the iOS app was now throwing errors with every message he sends me.

* How many times did this error show up in the conversation? - The recipient's key is not valid error pops up overtime he messages me on the iOS app, OS X app works fine.

* Please share a debug log from your device. (steps below) - https://gist.github.com/3144e7ad915345c0773d76b3ff06986d

**Comment by riyapenn** [comment]
New one:

* Was this 1:1 or group convo? 1:1 convo

* Did sender or receiver see these errors? receiver saw the errors. Sender (me) got no errors

* What Version of Signal are you using? I am using version 2.11.4

* What version of Signal is your contact on? Receiver is using version 2.11.4

* Had you or your contact recently used your phone or Signal before? Somewhat recently. But I broke my iPhone and had to get a new phone… this issue started after I got the new phone

* Had either phone not been used in a while? See answer above

* Had either of you reinstalled since you had last communicated? App was restored from a backup of my old/previous iPhone

* Did you see "safety numbers changed" ui right before these errors? Did these errors show up before or after you approved the new safety numbers? Were there more than one "safety numbers changed" message? No. No Safety Number messages

* Was there any other interaction on either party's part before these errors? No

* How many times did this error show up in the conversation? Each time I sent a message, the receiver sees the error

* Please share a debug log from your device. (steps below) https://gist.github.com/23143a2f1bcfaf9005830c5e03f244e6

* Please share a debug log from your contacts device as well. https://gist.github.com/c4eb846492473dcda8563a68585b056f

* Deleting the conversation thread on both devices did not fix the problem

**Comment by michaelkirk** [comment]
In the second case: https://github.com/WhisperSystems/Signal-iOS/issues/1838#issuecomment-301652747

The user had just updated from `2.10.1.2`, and had several keys that were incorrectly marked as unaccepted, so they were deleted upon launch. When they went to accept the changed SN and decrypt, they no longer had the necessary keying material.

This was partially fixed in 2.11, in that we're now correctly marking keys as accepted, but moreso in 2.12 which ensures we don't delete the last 3 keys, regardless of if they were marked as accepted.

There is still a potential issue wherein if you don't accept someones' SN change for a long time (like a week), you'll rotate out the necessary keys for decrypting. The solution for this should be in 2.13.

**Comment by riyapenn** [comment]
Triggered by receiver and sender. 
Alice's Log: https://gist.github.com/ce57a836d77ef3b33921c444f5eb018e

- This is a 1:1 convo. It only happens between Alice and Bob and not with any other contacts.
- Alice sees these errors 
- Alice and Bob are on Signal iOS 2.12.2
- Alice views the messages from the Chrome app on a laptop with no errors.
- Alice and Bob use Signal daily. 
- Bob had installed Signal onto a new phone and the issue occurred. However, the issue stopped for a couple of days, then began again. 
- Alice does not see "safety numbers changed" error prior to key not valid error.
- The error message appears each time Alice or Bob sends a message.

**Comment by AnirudhVyas** [comment]
removing a contact and adding it back makes the error go away. I am not sure why this was never fixed?

**Comment by michaelkirk-signal** [comment]
@AnirudhVyas  - The underlying problem to this issue was fixed long ago, so I'm going to close this issue.

A new problem could have the same symptom. If this has happened within the last three days, and you didn't disable debug logging, could you open a new issue and include your debug log and ideally the debug logger from the other party?

Also we'll need to know at what time the problematic message appeared so we can cross reference it in the debug log.

**Comment by AnirudhVyas** [comment]
This happened yesterday for me, deleted contact and added again and it worked ok - not sure why that happens in the first place

**Comment by michaelkirk-signal** [comment]
@AnirudhVyas - As I said, the underlying problem to this issue was fixed long ago.

A new problem could have the same symptom. If this has happened within the last three days, and you didn't disable debug logging, could you open a new issue and include your debug log and ideally the debug logger from the other party?

To follow up on your issue we'll need some details. Please file a new issue. I'm gong to lock this issue from discouraging you from continuing following up here, since it's not actionable without the details I've requested.
